In [30]:
import pandas

读入相关数据表，然后查看每行所代表的具体内容
---

In [31]:
data=pandas.read_csv('../data/data.csv',encoding='gbk')
data.head(5)

Unnamed: 0   custid                          trade_no bank_card_no  \
0           5  2791858  20180507115231274000000023057383          卡号1   
1          10   534047  20180507121002192000000023073000          卡号1   
2          12  2849787  20180507125159718000000023114911          卡号1   
3          13  1809708  20180507121358683000000388283484          卡号1   
4          14  2499829  20180507115448545000000388205844          卡号1   

   low_volume_percent  middle_volume_percent  \
0                0.01                   0.99   
1                0.02                   0.94   
2                0.04                   0.96   
3                0.00                   0.96   
4                0.01                   0.99   

   take_amount_in_later_12_month_highest  trans_amount_increase_rate_lately  \
0                                      0                               0.90   
1                                   2000                               1.28   
2                                      0                               1.00   
3                                   2000                               0.13   
4                                      0                               0.46   

   trans_activity_month  trans_activity_day  ...  loans_max_limit  \
0                  0.55               0.313  ...           2900.0   
1                  1.00               0.458  ...           3500.0   
2                  1.00               0.114  ...           1600.0   
3                  0.57               0.777  ...           3200.0   
4                  1.00               0.175  ...           2300.0   

   loans_avg_limit  consfin_credit_limit  consfin_credibility  \
0           1688.0                1200.0                 75.0   
1           1758.0               15100.0                 80.0   
2           1250.0                4200.0                 87.0   
3           1541.0               16300.0                 80.0   
4           1630.0                8300.0                 79.0   

   consfin_org_count_current  consfin_product_count  consfin_max_limit  \
0                        1.0                    2.0             1200.0   
1                        5.0                    6.0            22800.0   
2                        1.0                    1.0             4200.0   
3                        5.0                    5.0            30000.0   
4                        2.0                    2.0             8400.0   

   consfin_avg_limit  latest_query_day  loans_latest_day  
0             1200.0              12.0              18.0  
1             9360.0               4.0               2.0  
2             4200.0               2.0               6.0  
3            12180.0               2.0               4.0  
4             8250.0              22.0             120.0  

[5 rows x 90 columns]

**查看数据的大致信息，并且进行简单分析**

In [32]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4754 entries, 0 to 4753
Data columns (total 90 columns):
Unnamed: 0                                    4754 non-null int64
custid                                        4754 non-null int64
trade_no                                      4754 non-null object
bank_card_no                                  4754 non-null object
low_volume_percent                            4752 non-null float64
middle_volume_percent                         4752 non-null float64
take_amount_in_later_12_month_highest         4754 non-null int64
trans_amount_increase_rate_lately             4751 non-null float64
trans_activity_month                          4752 non-null float64
trans_activity_day                            4752 non-null float64
transd_mcc                                    4752 non-null float64
trans_days_interval_filter                    4746 non-null float64
trans_days_interval                           4752 non-null float64
regional_mobility

**简单分析**  
查看上面的数据可知，总共是有4754行，然后总共分为90列
经过粗略的观察，发现student_feature这一列的内容与其他的不是很相同，它只有1756 non-null float64，即这一列存在大量的缺失值，后期处理的时候
需要特别注意（或者是直接删除）。另外也存在一些列存在缺失值，后面建模的时候要处理
数据类型部分：
在这个data里面，数据类型分为三种，即：float64,int64,以及object这三种数据类型，在建模的时候需要将objetc类型进行转换

In [33]:
# 查看数据的属性
for i,name in enumerate(data.columns):
    name_sum = data[name].value_counts().shape[0]
    print("{:2}、{:50} The number of types of features is：{}"
          .format(i+1, name, name_sum))

 1、Unnamed: 0                                         The number of types of features is：4754
 2、custid                                             The number of types of features is：4754
 3、trade_no                                           The number of types of features is：4754
 4、bank_card_no                                       The number of types of features is：1
 5、low_volume_percent                                 The number of types of features is：40
 6、middle_volume_percent                              The number of types of features is：90
 7、take_amount_in_later_12_month_highest              The number of types of features is：166
 8、trans_amount_increase_rate_lately                  The number of types of features is：782
 9、trans_activity_month                               The number of types of features is：84
10、trans_activity_day                                 The number of types of features is：512
11、transd_mcc                                         The number of type

In [34]:
# 数据描述
data.describe()

Unnamed: 0        custid  low_volume_percent  middle_volume_percent  \
count   4754.000000  4.754000e+03         4752.000000            4752.000000   
mean    6008.414178  1.690993e+06            0.021806               0.901294   
std     3452.071428  1.034235e+06            0.041527               0.144856   
min        5.000000  1.140000e+02            0.000000               0.000000   
25%     3106.000000  7.593358e+05            0.010000               0.880000   
50%     6006.500000  1.634942e+06            0.010000               0.960000   
75%     8999.000000  2.597905e+06            0.020000               0.990000   
max    11992.000000  4.004694e+06            1.000000               1.000000   

       take_amount_in_later_12_month_highest  \
count                            4754.000000   
mean                             1940.197728   
std                              3923.971494   
min                                 0.000000   
25%                                 0.000000   
50%                               500.000000   
75%                              2000.000000   
max                             68000.000000   

       trans_amount_increase_rate_lately  trans_activity_month  \
count                        4751.000000           4752.000000   
mean                           14.160674              0.804411   
std                           694.180473              0.196920   
min                             0.000000              0.120000   
25%                             0.615000              0.670000   
50%                             0.970000              0.860000   
75%                             1.600000              1.000000   
max                         47596.740000              1.000000   

       trans_activity_day   transd_mcc  trans_days_interval_filter  ...  \
count         4752.000000  4752.000000                 4746.000000  ...   
mean             0.365425    17.502946                   29.029920  ...   
std              0.170196     4.475616                   22.722432  ...   
min              0.033000     2.000000                    0.000000  ...   
25%              0.233000    15.000000                   16.000000  ...   
50%              0.350000    17.000000                   23.000000  ...   
75%              0.480000    20.000000                   32.000000  ...   
max              0.941000    42.000000                  285.000000  ...   

       loans_max_limit  loans_avg_limit  consfin_credit_limit  \
count      4457.000000      4457.000000           4457.000000   
mean       3390.038142      1820.357864           9187.009199   
std        1474.206546       583.418291           7371.257043   
min           0.000000         0.000000              0.000000   
25%        2300.000000      1535.000000           4800.000000   
50%        3100.000000      1810.000000           7700.000000   
75%        4300.000000      2100.000000          11700.000000   
max       10000.000000      6900.000000          87100.000000   

       consfin_credibility  consfin_org_count_current  consfin_product_count  \
count          4457.000000                4457.000000            4457.000000   
mean             76.042630                   4.732331               5.227507   
std              14.536819                   2.974596               3.409292   
min               0.000000                   0.000000               0.000000   
25%              77.000000                   2.000000               3.000000   
50%              79.000000                   4.000000               5.000000   
75%              80.000000                   7.000000               7.000000   
max              87.000000                  18.000000              20.000000   

       consfin_max_limit  consfin_avg_limit  latest_query_day  \
count        4457.000000        4457.000000       4450.000000   
mean        16153.690823        8007.696881         24.112809   
std         14301.037628        5679.418585         37.725724   
min             0

**describe分析**  
首先，这是统计出来了83列的信息，即数据中object类型存在7列（不能用计数、均值等等去描述），然后在前面可知：int64 共有13列，float64 共有70列
然后更深层次的信息分析：  
大部分数据特征存在拖尾现象， 即特征的 max(min) 明显偏离其 mean ，需要对这些数据进行进一步的处理才能建立准确的模型。
再是结合上面的名称以及在表中均值方差的特点，可以看到部分列的数据类型存在问题，如'first_transaction_time'是data类型而不是float64 类型

无关特征的删除
----
分析data，有些数据对于建立模型是没有任何用处的,可以结合一下具体的贷款场景以及上面中的number of feature
- Unnamed: 0'
- 客户id：'custid'
- 流水号：'trade_no'
- 卡号：'bank_card_no'
- 资源：'source'
- 客户姓名：'id_name'
- 最新查询时间：'latest_query_time'
- 最新贷款时间：'loans_latest_time'  
这些特征也可以进行删除

In [35]:
data=data.drop(['Unnamed: 0','custid','trade_no','source','id_name','latest_query_time','loans_latest_time','bank_card_no'],axis=1)


In [37]:
# 删除了8列之后的数据变成了下面这样的形式
data.head(5)

low_volume_percent  middle_volume_percent  \
0                0.01                   0.99   
1                0.02                   0.94   
2                0.04                   0.96   
3                0.00                   0.96   
4                0.01                   0.99   

   take_amount_in_later_12_month_highest  trans_amount_increase_rate_lately  \
0                                      0                               0.90   
1                                   2000                               1.28   
2                                      0                               1.00   
3                                   2000                               0.13   
4                                      0                               0.46   

   trans_activity_month  trans_activity_day  transd_mcc  \
0                  0.55               0.313        17.0   
1                  1.00               0.458        19.0   
2                  1.00               0.114        13.0   
3                  0.57               0.777        22.0   
4                  1.00               0.175        13.0   

   trans_days_interval_filter  trans_days_interval  regional_mobility  ...  \
0                        27.0                 26.0                3.0  ...   
1                        30.0                 14.0                4.0  ...   
2                        68.0                 22.0                1.0  ...   
3                        14.0                  6.0                3.0  ...   
4                        66.0                 42.0                1.0  ...   

   loans_max_limit  loans_avg_limit  consfin_credit_limit  \
0           2900.0           1688.0                1200.0   
1           3500.0           1758.0               15100.0   
2           1600.0           1250.0                4200.0   
3           3200.0           1541.0               16300.0   
4           2300.0           1630.0                8300.0   

   consfin_credibility  consfin_org_count_current  consfin_product_count  \
0                 75.0                        1.0                    2.0   
1                 80.0                        5.0                    6.0   
2                 87.0                        1.0                    1.0   
3                 80.0                        5.0                    5.0   
4                 79.0                        2.0                    2.0   

   consfin_max_limit  consfin_avg_limit  latest_query_day  loans_latest_day  
0             1200.0             1200.0              12.0              18.0  
1            22800.0             9360.0               4.0               2.0  
2             4200.0             4200.0               2.0               6.0  
3            30000.0            12180.0               2.0               4.0  
4             8400.0             8250.0              22.0             120.0  

[5 rows x 82 columns]

数据类型的转换
---
在前面的分析中可知，部分的列的数据类型与实际的数据类型不符合，另外，不是所有的数据类型都是数值类型，需要先转换为数值类型


In [38]:
data.reg_preference_for_trad.unique()

array(['一线城市', '三线城市', '境外', '二线城市', '其他城市', nan], dtype=object)

In [39]:
# 数据类型进行转换，将其进行映射
for i in range(0, data.shape[0]):
    if data.reg_preference_for_trad[i] == '一线城市':
        data.reg_preference_for_trad[i] = 1
    elif data.reg_preference_for_trad[i] == '二线城市':
        data.reg_preference_for_trad[i] = 2
    elif data.reg_preference_for_trad[i] == '三线城市':
        data.reg_preference_for_trad[i] = 3
    elif data.reg_preference_for_trad[i] == '境外':
        data.reg_preference_for_trad[i] = 4
    elif data.reg_preference_for_trad[i] == '其他城市':
        data.reg_preference_for_trad[i] = 5

J:\python\env\nlp\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.
J:\python\env\nlp\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
J:\python\env\nlp\lib\site-packages\ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # Remove the CWD from sys.path while we load stuff.
J:\python\env\nlp\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyW

In [40]:
# 将其他的object类型进行转换

data = data.convert_objects(convert_numeric=True)
data.dtypes.value_counts()

J:\python\env\nlp\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: convert_objects is deprecated.  To re-infer data dtypes for object columns, use DataFrame.infer_objects()
For all other conversions use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.
  This is separate from the ipykernel package so we can avoid doing imports until


float64    71
int64      11
dtype: int64

缺失值处理
----
在这里，主要是学习缺失值的处理规则，什么样的列中缺失值应该丢弃，什么样的缺失值应该进行填补，采用什么方式进行填补
【PS】有的可以采用简单的用均值进行填充，用最大值、最小值进行填充，同样的，现在也是可以对某个值采用更加精细的，使用机器学习方法如决策树来进行填充缺失值  
- 删除缺失值较多的行列  
  若某个特征的缺失值超过 30%，那么它会损失特征的关键信息。  
  同样若某个样本的缺失值过多，那么它也不再具有统计学意义。  

因此我们对这样的行或列直接进行删除。

In [41]:
data = data.dropna(axis=1, thresh = 1000)#删除缺失值超过 70% 的列
data = data.dropna(axis=0, thresh = 75)#删除缺失值超过7个的行

- 填充缺失值

对于剩余的缺失值，这里采用 平均值填充法 进行填充。

In [43]:
# 注意代码编写方式
data = data.fillna(data.mean())
data.head()

low_volume_percent  middle_volume_percent  \
0                0.01                   0.99   
1                0.02                   0.94   
2                0.04                   0.96   
3                0.00                   0.96   
4                0.01                   0.99   

   take_amount_in_later_12_month_highest  trans_amount_increase_rate_lately  \
0                                      0                               0.90   
1                                   2000                               1.28   
2                                      0                               1.00   
3                                   2000                               0.13   
4                                      0                               0.46   

   trans_activity_month  trans_activity_day  transd_mcc  \
0                  0.55               0.313        17.0   
1                  1.00               0.458        19.0   
2                  1.00               0.114        13.0   
3                  0.57               0.777        22.0   
4                  1.00               0.175        13.0   

   trans_days_interval_filter  trans_days_interval  regional_mobility  ...  \
0                        27.0                 26.0                3.0  ...   
1                        30.0                 14.0                4.0  ...   
2                        68.0                 22.0                1.0  ...   
3                        14.0                  6.0                3.0  ...   
4                        66.0                 42.0                1.0  ...   

   loans_max_limit  loans_avg_limit  consfin_credit_limit  \
0           2900.0           1688.0                1200.0   
1           3500.0           1758.0               15100.0   
2           1600.0           1250.0                4200.0   
3           3200.0           1541.0               16300.0   
4           2300.0           1630.0                8300.0   

   consfin_credibility  consfin_org_count_current  consfin_product_count  \
0                 75.0                        1.0                    2.0   
1                 80.0                        5.0                    6.0   
2                 87.0                        1.0                    1.0   
3                 80.0                        5.0                    5.0   
4                 79.0                        2.0                    2.0   

   consfin_max_limit  consfin_avg_limit  latest_query_day  loans_latest_day  
0             1200.0             1200.0              12.0              18.0  
1            22800.0             9360.0               4.0               2.0  
2             4200.0             4200.0               2.0               6.0  
3            30000.0            12180.0               2.0               4.0  
4             8400.0             8250.0              22.0             120.0  

[5 rows x 82 columns]

数据集切分
----
对于数据集切分，我们可以直接进行切分，也可以借助于sklearn进行处理

In [48]:
# 采用基础的方法进行切分数据集,
from sklearn.utils import shuffle
data=shuffle(data)

train_data_len=int(len(data)*0.7)
train_data=data[:train_data_len]
test_data=data[train_data_len+1:]

In [49]:
# 直接采用sklearn 里面的包来进行训练集、测试集切分
from sklearn.model_selection import train_test_split

train_data, test_data = train_test_split(data, test_size=0.3, random_state=2018)

In [50]:
# 将训练集和测试集的数据进行保存
train_data.to_csv('../data/train_data.csv', index=False, header=True)
test_data.to_csv('../data/test_data.csv', index=False, header=True)